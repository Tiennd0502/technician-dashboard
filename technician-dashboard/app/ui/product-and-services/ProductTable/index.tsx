'use client';

import React, { useCallback, useMemo, useState } from 'react';
import { Box, Button, Flex, Heading, VStack, useDisclosure, useToast } from '@chakra-ui/react';

// Types
import { Filter, Product, STATUS, TableData } from '@/lib/interfaces';

// Constants
import { DEFAULT_PRODUCT_FILTER, ROUTES } from '@/lib/constants';

// Utils
import { removeSpaces } from '@/lib/utils';

// Components / icons
import { ConfirmModal, Pagination, SearchBox, Spinner, StatusLabel, Table } from '@/ui/commons';
import { FilterIcon } from '@/ui/icons';
import ProductForm from '../ProductForm';

// Hooks
import { useFetchProducts, useProduct } from '@/lib/hooks';

const ProductTable = () => {
  const toast = useToast();

  const [productEdit, setProductEdit] = useState<Product>();
  const [productFilter, setProductFilter] = useState<Filter>(DEFAULT_PRODUCT_FILTER);
  const [productDelete, setProductDelete] = useState<string>('');

  const { isOpen: isOpenForm, onToggle: onToggleForm } = useDisclosure();
  const {
    isOpen: isOpenConfirmModal,
    onOpen: onOpenConfirmModal,
    onClose: onCloseConfirmModal,
  } = useDisclosure();

  const {
    deleteProduct: { mutate: deleteProduct, isPending: isDeleting },
  } = useProduct();

  const { data: products = [], isLoading } = useFetchProducts(productFilter);

  // TODO: Config api return total pages
  const { data: productAll = [] } = useFetchProducts({
    name: productFilter.name || '',
    sortBy: productFilter.sortBy,
    order: productFilter.order,
  });

  const totalPage = useMemo(
    () => Math.ceil(productAll.length / +DEFAULT_PRODUCT_FILTER.limit),
    [productAll],
  );

  const handleCloseForm = useCallback(() => {
    productEdit && setProductEdit(undefined);
    onToggleForm();
  }, [onToggleForm, productEdit]);

  const handleSortProduct = useCallback((value: Filter) => {
    setProductFilter((prev: Filter) => ({
      ...prev,
      ...value,
    }));
  }, []);

  const handleClickEditProduct = useCallback(
    (id: string) => {
      setProductEdit(products?.find((product) => product.id === id));
      onToggleForm();
    },
    [onToggleForm, products],
  );

  const handleSearchProduct = useCallback((value: string) => {
    setProductFilter((prev: Filter) => ({
      ...prev,
      name: removeSpaces(value, true),
      page: '1',
    }));
  }, []);

  const handleCloseConfirmModal = useCallback(() => {
    setProductDelete('');
    onCloseConfirmModal();
  }, [onCloseConfirmModal]);

  const handleDeleteProduct = useCallback(() => {
    deleteProduct(productDelete, {
      onSuccess: () => {
        handleCloseConfirmModal();
        toast({
          position: 'top-right',
          title: `Product deleted.`,
          status: 'success',
          duration: 9000,
          isClosable: true,
        });
      },
      onError: () => {
        toast({
          position: 'top-right',
          title: `Product delete failed.`,
          status: 'error',
          duration: 9000,
          isClosable: true,
        });
      },
    });
  }, [deleteProduct, handleCloseConfirmModal, productDelete, toast]);

  const handleOpenConfirmModal = useCallback(
    (id: string) => {
      setProductDelete(id);
      onOpenConfirmModal();
    },
    [onOpenConfirmModal],
  );

  const handleChangePage = useCallback((page: number) => {
    setProductFilter((prev: Filter) => ({
      ...prev,
      page: page.toString(),
    }));
  }, []);

  const productHeaderColumn = useMemo(() => {
    const customViewStatus = (value: string | number | boolean) => (
      <StatusLabel value={value as STATUS} />
    );

    return [
      {
        key: 'checkbox',
        label: 'Checkbox',
        width: {
          md: '30px',
          lg: '50px',
        },
        isCheckbox: true,
      },
      {
        key: 'name',
        label: 'Product name',
        width: {
          md: '100%',
          lg: '28%',
        },
        onSort: handleSortProduct,
      },
      {
        key: 'brand',
        label: 'Brand name',
        width: {
          md: '100%',
          lg: '17%',
        },
      },
      {
        key: 'service',
        label: 'Service',
        width: {
          md: '100%',
          lg: '15%',
        },
      },
      {
        key: 'status',
        label: 'Status',
        width: {
          md: '100%',
          lg: '18%',
        },
        customView: customViewStatus,
      },
      {
        key: 'actions',
        label: 'Actions',
        width: {
          md: '100%',
          lg: 'calc(32% - 35px)',
        },
        isAction: true,
      },
    ];
  }, [handleSortProduct]);

  return (
    <VStack spacing={7} p='7' borderWidth='1px' borderRadius='md' borderColor='primary' h='100%'>
      {isOpenForm && (
        <ProductForm
          title={`${productEdit ? 'Edit' : 'Add'} Product`}
          isOpen={isOpenForm}
          product={productEdit}
          onClose={handleCloseForm}
        />
      )}
      {isOpenConfirmModal && (
        <ConfirmModal
          title='Delete Product'
          description='Are you sure you want to delete this product?'
          isOpen={isOpenConfirmModal}
          isSubmitting={isDeleting}
          onSubmit={handleDeleteProduct}
          onClose={handleCloseConfirmModal}
        />
      )}
      <Flex
        gap={{ base: 4, md: 2, lg: 0 }}
        justifyContent='space-between'
        alignItems='center'
        w='full'
        flexWrap='wrap'
      >
        <Heading variant='headingLg' mr='auto' w={{ base: '100%', lg: '25%' }}>
          Products listing
        </Heading>
        <Box minW='200px' w={{ base: '100%', md: '50%', lg: '40%' }}>
          <SearchBox onSearch={handleSearchProduct} />
        </Box>
        <Button as={Flex} gap='4' variant='outline' ml={{ md: 'auto' }} mr={{ md: 4 }}>
          Filter <FilterIcon />
        </Button>
        <Button onClick={onToggleForm}>Add new</Button>
      </Flex>

      {isLoading ? (
        <Flex minH='388px' alignItems='center'>
          <Spinner />
        </Flex>
      ) : (
        <>
          <Table
            filter={productFilter}
            columns={productHeaderColumn}
            data={products as unknown as TableData[]}
            onEdit={handleClickEditProduct}
            onDelete={handleOpenConfirmModal}
          />

          {!!products.length && (
            <Pagination
              onChange={handleChangePage}
              total={totalPage}
              page={+(productFilter?.page || 1)}
            />
          )}
        </>
      )}
    </VStack>
  );
};

export default ProductTable;
