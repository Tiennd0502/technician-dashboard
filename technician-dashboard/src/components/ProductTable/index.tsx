'use client';

import React, { useCallback, useMemo, useState } from 'react';
import { Box, useDisclosure } from '@chakra-ui/react';

// Types
import { Filter, STATUS, TableData } from '@/interfaces';

// Constants
import { DEFAULT_PRODUCT_FILTER } from '@/constants';

// Components
import { StatusLabel, Table } from '..';

// Hooks
import { useFetchProducts } from '@/hooks';

const ProductTable = () => {
  const [productFilter, setProductFilter] = useState<Filter>(DEFAULT_PRODUCT_FILTER);

  const { isOpen: isOpenForm, onOpen: onOpenForm, onClose: onCloseForm } = useDisclosure();
  const {
    isOpen: isOpenConfirmModal,
    onOpen: onOpenConfirmModal,
    onClose: onCloseConfirmModal,
  } = useDisclosure();

  const { data: products = [] } = useFetchProducts(DEFAULT_PRODUCT_FILTER);

  const handleSortProduct = useCallback((value: Filter) => {
    setProductFilter((prev: Filter) => ({
      ...prev,
      ...value,
    }));
  }, []);

  const handleClickEditProduct = useCallback((id: string) => {}, []);

  const handleSearchProduct = useCallback((value: string) => {
    setProductFilter((prev: Filter) => ({
      ...prev,
      name: value,
    }));
  }, []);

  const handleOpenConfirmModal = useCallback(
    (id: string) => {
      onOpenConfirmModal();
    },
    [onOpenConfirmModal],
  );

  const productHeaderColumn = useMemo(() => {
    const customViewStatus = (value: string | number | boolean) => (
      <StatusLabel value={value as STATUS} />
    );

    return [
      {
        key: 'checkbox',
        label: 'Checkbox',
        width: {
          md: '30px',
          lg: '50px',
        },
        isCheckbox: true,
      },
      {
        key: 'name',
        label: 'Product name',
        width: {
          md: '100%',
          lg: '28%',
        },
        onSort: handleSortProduct,
      },
      {
        key: 'brand',
        label: 'Brand name',
        width: {
          md: '100%',
          lg: '17%',
        },
      },
      {
        key: 'service',
        label: 'Service',
        width: {
          md: '100%',
          lg: '15%',
        },
      },
      {
        key: 'status',
        label: 'Status',
        width: {
          md: '100%',
          lg: '18%',
        },
        customView: customViewStatus,
      },
      {
        key: 'actions',
        label: 'Actions',
        width: {
          md: '100%',
          lg: 'calc(32% - 35px)',
        },
        isAction: true,
      },
    ];
  }, [handleSortProduct]);

  return (
    <Box>
      <Table
        title='Products listing'
        filter={productFilter}
        columns={productHeaderColumn}
        data={products as unknown as TableData[]}
        onAdd={onOpenForm}
        onEdit={handleClickEditProduct}
        onDelete={handleOpenConfirmModal}
        onSearch={handleSearchProduct}
      />
    </Box>
  );
};

export default ProductTable;
