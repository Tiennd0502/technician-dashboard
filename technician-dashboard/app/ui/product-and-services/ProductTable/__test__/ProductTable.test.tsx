import { render } from '@testing-library/react';

import * as ReactQuery from '@tanstack/react-query';
import { ChakraProvider } from '@chakra-ui/react';

import ProductTable from '..';

const queryClient = new ReactQuery.QueryClient({
  defaultOptions: {
    queries: {
      retry: false,
    },
  },
});

const mockMutateAsync = jest.fn();

jest.mock('@/lib/hooks', () => ({
  ...jest.requireActual('@/lib/hooks'),
  useProduct: () => ({
    deleteProduct: {
      mutateAsync: mockMutateAsync,
      isLoading: false,
    },
  }),
}));

const setup = () =>
  render(
    <ChakraProvider>
      <ReactQuery.QueryClientProvider client={queryClient}>
        <ProductTable />
      </ReactQuery.QueryClientProvider>
    </ChakraProvider>,
  );

describe('Product Table test cases', () => {
  test('should render correctly', async () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });
});
