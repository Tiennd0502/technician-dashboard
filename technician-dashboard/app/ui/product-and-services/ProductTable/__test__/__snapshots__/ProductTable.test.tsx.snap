// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Table test cases should render correctly 1`] = `
<div>
  <div
    class="chakra-stack css-1sawn9g"
  >
    <div
      class="css-c7dwok"
    >
      <h2
        class="chakra-heading css-huystf"
      >
        Products listing
      </h2>
      <div
        class="css-12olhs5"
      >
        <div
          class="chakra-input__group css-1y0e7gb"
          data-group="true"
        >
          <input
            class="chakra-input css-scsyde"
            placeholder="Enter keyword..."
            value=""
          />
          <div
            class="chakra-input__right-element css-1lds0jh"
          >
            <svg
              class="chakra-icon css-w4ub0m"
              data-testid="search-icon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M23.384,21.619,16.855,15.09a9.284,9.284,0,1,0-1.768,1.768l6.529,6.529a1.266,1.266,0,0,0,1.768,0A1.251,1.251,0,0,0,23.384,21.619ZM2.75,9.5a6.75,6.75,0,1,1,6.75,6.75A6.758,6.758,0,0,1,2.75,9.5Z"
                fill="currentColor"
              />
            </svg>
          </div>
        </div>
      </div>
      <div
        class="chakra-button css-1ti8ltj"
      >
        Filter 
        <svg
          fill="none"
          height="18"
          viewBox="0 0 16 18"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M10.0571 8.49269H15.4857C15.7333 8.49269 15.9619 8.72684 15.9619 8.98049C15.9619 9.25366 15.7524 9.4683 15.4857 9.4683H10.0571C9.8476 10.4634 8.97141 11.2049 7.94284 11.2049C6.91427 11.2049 6.03808 10.4634 5.82856 9.4683H0.533319C0.266653 9.4683 0.0571289 9.25366 0.0571289 8.98049C0.0571289 8.70732 0.266653 8.49269 0.533319 8.49269H5.82856C6.03808 7.49757 6.91427 6.7561 7.94284 6.7561C8.97141 6.7561 9.8476 7.49757 10.0571 8.49269ZM6.7238 8.98049C6.7238 9.66342 7.27618 10.2293 7.94284 10.2293C8.60951 10.2293 9.16189 9.66342 9.16189 8.98049C9.16189 8.29757 8.60951 7.73171 7.94284 7.73171C7.27618 7.73171 6.7238 8.29757 6.7238 8.98049Z"
            fill="#222"
            fill-rule="evenodd"
          />
          <path
            clip-rule="evenodd"
            d="M4.36192 3.82938C4.15239 4.8245 3.2762 5.56597 2.24763 5.56597C1.04763 5.56597 0.0762024 4.55133 0.0762024 3.34158C0.0762024 2.13182 1.04763 1.11719 2.24763 1.11719C3.2762 1.11719 4.15239 1.85865 4.36192 2.85377H15.4667C15.7333 2.85377 15.9429 3.06841 15.9429 3.34158C15.9429 3.61475 15.7333 3.82938 15.4667 3.82938H4.36192ZM3.46668 3.34158C3.46668 2.65865 2.9143 2.0928 2.24763 2.0928C1.58096 2.0928 1.02858 2.65865 1.02858 3.34158C1.02858 4.0245 1.58096 4.59036 2.24763 4.59036C2.9143 4.59036 3.46668 4.0245 3.46668 3.34158Z"
            fill="#222"
            fill-rule="evenodd"
          />
          <path
            clip-rule="evenodd"
            d="M11.6571 14.1512C11.8857 13.1755 12.7428 12.4341 13.7714 12.4341C14.9524 12.4341 15.9428 13.4292 15.9428 14.6585C15.9428 15.8877 14.9714 16.8829 13.7714 16.8829C12.7428 16.8829 11.8667 16.1219 11.6571 15.1268H0.533319C0.266653 15.1268 0.0571289 14.9121 0.0571289 14.639C0.0571289 14.3658 0.266653 14.1512 0.533319 14.1512H11.6571ZM12.5524 14.639C12.5524 15.3219 13.1047 15.8877 13.7714 15.8877C14.4381 15.8877 14.9905 15.3219 14.9905 14.639C14.9905 13.956 14.4381 13.3902 13.7714 13.3902C13.1047 13.3902 12.5524 13.956 12.5524 14.639Z"
            fill="#222"
            fill-rule="evenodd"
          />
        </svg>
      </div>
      <button
        class="chakra-button css-ez23ye"
        type="button"
      >
        Add new
      </button>
    </div>
    <div
      class="css-bclwcc"
    >
      <div
        class="css-1n2mv2k"
      >
        <div
          class="chakra-spinner css-1gumy0q"
        >
          <span
            class="css-8b45rq"
          >
            Loading...
          </span>
        </div>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
